File: Python/main_encoder.py
import numpy as np
import cv2 
from PIL import Image
import matplotlib.pyplot as plt
from crossset import crossset
from dotset import dotset
from embeded_modification2 import embeded
np.set_printoptions(threshold=np.inf)
# Import an image.
original_image = Image.open('lenanew.tiff')
image_array = np.array(original_image)
# Divide RGB plane.
R = image_array[:, :, 0]
G = image_array[:, :, 1]  
B = image_array[:, :, 2]  
R = R.astype(float)
G = G.astype(float)
B = B.astype(float)
currentImage = R
cross_data = crossset(currentImage)
# Convert image to grayscale.
watermark = Image.open('w.bmp').convert("L")
threshold = 0.8 * 255  
watermark = watermark.point(lambda p: 1.0 if p > threshold else 0.0)
watermark = np.array(watermark)
# Convert datatype to double.
watermark = watermark.astype(np.float64)  
# A column matrix.
watermark = watermark.flatten()
# Separate half watermark.
half_length = len(watermark) // 2
first_half = []
second_half = []
for i in range(len(watermark)):
    if i < half_length:
        first_half.append(watermark[i])
    else:
        second_half.append(watermark[i])
# Convert to array.
Pload_cross = np.array(first_half)
Pload_dot = np.array(second_half)
# Set.
max_kb = 26
maxT = 24
Tn = np.repeat(-np.arange(1, maxT+1), 2).reshape(-1, 1)
Tp = np.repeat(np.arange(1, maxT), 2).reshape(-1, 1)
TpTnpsnr = np.zeros((len(Tn), 3))
TpTnpsnr[:, 0] = np.concatenate((np.array([0]), Tp.flatten(), np.array([maxT])))
TpTnpsnr[:, 1] = Tn.flatten()
# Opimize value. 
oypsnr = np.max(TpTnpsnr[:, 2])
Id = np.argmax(TpTnpsnr[:, 2])  
embeded_cross_image, PLcheckcross = embeded(cross_data, Pload_cross, TpTnpsnr[Id, 0], TpTnpsnr[Id, 1], currentImage)
dot_data = dotset(embeded_cross_image)
embeded_dot_image, PLcheckdot = embeded(dot_data, Pload_dot, TpTnpsnr[Id, 0], TpTnpsnr[Id, 1], embeded_cross_image)
R = embeded_dot_image
# Complie RGB plane.
R = np.clip(R, 0, 255)
G = np.clip(G, 0, 255)
B = np.clip(B, 0, 255)
np.savetxt('image_array.txt', G.reshape(-1, G.shape[-1]), fmt='%d')
img_reconstructed = np.stack([B, G, R], axis=-1).astype(np.uint8)
cv2.imwrite('reconstructed_image.png', img_reconstructed)

File: MathLab/Text/main_encoder.m
clc;
close all;
clear all;
% Load an original image.
originalImage = imread('lenanew.tiff');
figure(1);
imshow(originalImage);
title('Original Image');
% Sperate an original image to RGB plane.
R = double(originalImage(:,:,1));
G = double(originalImage(:,:,2));
B = double(originalImage(:,:,3));
% Read text from file.
originalText = fileread('text_file.txt');

% Convert text to binary.
watermark = textToBinary(originalText); 
watermark = watermark(:)';
Pload_cross = watermark(1:length(watermark)/2);
Pload_dot = watermark(length(watermark)/2+1:end);
% Define parameter.
max_kb = 26;
maxT = 24;
Tn = [1;1]*(-1:-1:-maxT);
Tp = [1;1]*(1:maxT-1);
TpTnpsnr = zeros(numel(Tn),3);
TpTnpsnr(:,1:2) = [[0; Tp(:); maxT] Tn(:)];
% Proccessing a plane.
currentImage = R;
cross_data = crossset(currentImage);

for ii = 1:size(TpTnpsnr,1)
    Tp = TpTnpsnr(ii,1);
    Tn = TpTnpsnr(ii,2);
    [embeded_cross_image, PLcheckcross] = embeded_modification2(cross_data, Pload_cross, Tp, Tn, currentImage);
    
    dot_data = dotset(embeded_cross_image);
    [embeded_dot_image, PLcheckdot] = embeded_modification2(dot_data, Pload_dot, Tp, Tn, embeded_cross_image);
    
    if PLcheckcross * PLcheckdot == 1
        Mean2err = sum(sum((embeded_dot_image - currentImage).^2)) / (numel(currentImage));
        sdf = 255^2 / Mean2err;
        PSNR = 10 * log10(sdf);
        TpTnpsnr(ii,3) = PSNR;
        ypsnr = PSNR;
        xbps = length(watermark) / numel(currentImage);
        figure(3);
        mapp = colormap(lines(size(TpTnpsnr,1)));
        figure(4);
        plot(xbps, ypsnr, '.-', 'Color', mapp(ii,:)), axis([0 1 25 60]), hold on, grid on;
        title(['Tp = ', num2str(Tp), ' Tn = ', num2str(Tn)]);
    end
    
    if sum(TpTnpsnr(:,3) > 0) > 2
        break;
    end
end

% Optimize value.
[oypsnr, Id] = max(TpTnpsnr(:,3));
[embeded_cross_image, PLcheckcross] = embeded_modification2(cross_data, Pload_cross, TpTnpsnr(Id,1), TpTnpsnr(Id,2), currentImage);
dot_data = dotset(embeded_cross_image);
[embeded_dot_image, PLcheckdot] = embeded_modification2(dot_data, Pload_dot, TpTnpsnr(Id,1), TpTnpsnr(Id,2), embeded_cross_image);
R = embeded_dot_image;
% Complie 3 plane to image RGB.
embeded = cat(3, R, G, B);
embeded = uint8(embeded);
imwrite(embeded, 'lena512_encoder.bmp');
figure(5);
imshow(embeded);
title('Encryption Image');

function data = textToBinary(text)
    % Convert text to uint8.
    byteData = uint8(text);
    % Convert uint8 to binary.
    % Binry array.    
    data = false(1, numel(byteData) * 8); 
    for i = 1:length(byteData)
        % Convert each uint8 to binary 8 bits.
        bits = dec2bin(byteData(i), 8) - '0';  
        % Collect in binary array.        
        data((i-1)*8+1:i*8) = logical(bits);  
    end
end


